### PR TITLE
Support configurable REST catalog base URI (Lakekeeper, Nessie, Unity)

### PR DESCRIPTION
**Pull request by sfc-gh-rbachala**
## Problem

The REST catalog URL templates hardcoded Polaris's `/api/catalog` mount
prefix, so pg_lake could only reach Polaris-style endpoints.  Iceberg REST
catalogs that mount the spec at a different path were unreachable:

| Catalog        | Mount path                              |
|----------------|-----------------------------------------|
| Polaris        | `/api/catalog`                          |
| Lakekeeper     | `/catalog`                              |
| Nessie         | `/iceberg`                              |
| Unity Catalog  | `/api/2.1/unity-catalog/iceberg`        |

## Solution

Drop `/api/catalog` from the `REST_CATALOG_*` URL templates so they carry
only the `/v1/...` Iceberg REST spec path.  `ResolveRestCatalogBaseUri` now
just strips one trailing slash and returns the endpoint verbatim — the full
mount path must be included in `rest_endpoint`.

The struct field `host` is renamed to `baseUri` to reflect that it now holds
an arbitrary base URI with a path, not just a host.

Normalization runs once in both option builders
(`BuildRestCatalogOptionsFromServer` and
`BuildRestCatalogOptionsFromUserMapping`), covering both the read path and the
write path (transaction-commit URL builder).  DuckDB receives only the resolved
S3 metadata location, so no DuckDB-side change is needed.

Also removes two duplicate `REST_CATALOG_AUTH_TOKEN_PATH` defines and the
unused `GET_REST_CATALOG_METADATA_LOCATION` macro.

**Breaking change (3.5):** `rest_endpoint` must now include the full mount path.
Polaris example: `https://host/api/catalog`.  The GUC
`pg_lake_iceberg.rest_catalog_host` default changes from
`http://localhost:8181` to `http://localhost:8181/api/catalog`.
Existing bare-host configs must be updated.

## Test plan

- [x] **Unit test** — `pg_lake_iceberg/tests/pytests/test_rest_catalog_base_uri.py`
  — parametrized cases covering: explicit mount paths for Polaris/Lakekeeper/Nessie/Unity,
  trailing-slash variants, empty string.  SQL-callable wrapper exposes
  `ResolveRestCatalogBaseUri` so these assertions run without a live catalog.

- [x] **Integration test** — `test_rest_endpoint_path_is_honored[explicit]` in
  `pg_lake_table/tests/pytests/test_modify_iceberg_rest_table.py`
  — full round-trip (CREATE SERVER + USER MAPPING + CREATE TABLE + INSERT 10
  rows + SELECT count) against a live Polaris instance with an explicit mount path.

- [x] **Regression** — existing Polaris suite in `test_modify_iceberg_rest_table.py`
  unchanged.

## Notes for reviewers

- `oauthHostPath` (the `oauth_endpoint` server option) is treated as a
  fully-qualified URL and intentionally **not** passed through
  `ResolveRestCatalogBaseUri` — it is often a separate IdP host.

## Follow-ups (to be filed separately)

- `/v1/config` endpoint discovery to auto-detect the catalog prefix
- Optional `client_id` for unauthenticated Nessie servers
- Live Nessie and Lakekeeper fixtures in the integration test suite

### CODE REVIEW

**Comment by sfc-gh-mslot** [comment]
Pushed 48f4467 rebased onto current main as `marcoslot/ci-pr-507` so the full suite actually runs (fork PRs don't trigger `pytest_all`): https://github.com/Snowflake-Labs/pg_lake/actions/runs/31593237833

Builds are green on pg16/17/18 on both almalinux and debian. All 10 `test_resolve_rest_catalog_base_uri` cases pass, and `test_rest_endpoint_path_is_honored[bare]`/`[explicit]` pass on 17 and 18 against live Polaris. `test-installcheck (16, installcheck-pg_lake_spatial)` failed, but that one is unrelated: pgduck_server never came up in that runner, so every spatial test died on `connection to server on socket "/tmp/.s.PGSQL.5332" failed`. Same job passes on main.

lint is a real failure though. pgindent wants to reflow the three new comment blocks, `make reindent` fixes it. The trailing comment needs the standard alignment too:

```
-	char	   *host;	/* normalized base URI; see ResolveRestCatalogBaseUri */
+	char	   *host;			/* normalized base URI; see
+								 * ResolveRestCatalogBaseUri */
```

On the design: I think a catalog mounted at the root is still unreachable. `https://host` and `https://host/` both get `/api/catalog` appended, so there's no way to say "the spec is at /v1 directly". The iceberg-rest-fixture reference image serves exactly that, and Lakekeeper can be deployed at the root as well. Do we want an explicit way to opt out?

One option is to strip the trailing slash only *after* the path check, so `https://host/` means root and `https://host` keeps the legacy prefix. That's fairly subtle, so maybe a separate option that can be set to empty is clearer. Probably not blocking, but it seems a shame to ship "configurable base URI" with the root case still missing.

Also, both parametrizations of the integration test end up at the exact same URL (`http://localhost:PORT/api/catalog/v1/...`), so it proves the verbatim path is honored but not that a *different* prefix works end to end. Fine given the follow-ups you list, just noting the coverage is a bit weaker than it reads.

Two nits: `host` now holds a base URI with a path in it, so `host -> baseUri` would read better at the psprintf sites. And the `if (baseUri == NULL) PG_RETURN_NULL()` in the test wrapper is dead code since the SQL function is declared STRICT.

Otherwise this looks right to me. Nothing else in the tree builds a REST URL outside the macros, `opts->host` is only read in those psprintf calls, nothing hands the catalog URI to duckdb, and normalizing in both option builders does cover read and write. `ResolveRestCatalogBaseUri` is also idempotent, so ALTER SERVER round trips won't compound the prefix.

**Comment by sfc-gh-rbachala** [comment]
Thanks for the full run and the thorough review.

**Lint / pgindent**
Fixed: `*host` field comment now uses three tabs before `/*` (aligning with the other `char *` fields at col 32 with 4-char tabs) and wraps to a second line.

**host → baseUri rename**
Done across the struct definition, `rest_catalog.c`, `rest_catalog_auth.c`, `rest_catalog_ops.c`, `rest_catalog_options.c`, and the URL-templates comment in the header.  Prose mentions of "bare-host" and "host[:port]" in block comments are left as-is since they describe endpoint patterns, not the field.

**Dead code in test wrapper**
Removed: `if (baseUri == NULL) PG_RETURN_NULL()` folded away; wrapper is now a single-line `PG_RETURN_TEXT_P(cstring_to_text(ResolveRestCatalogBaseUri(endpoint)))`.

**Root-mount case (https://host/)**
Went with your option 1 (check-before-strip): the `strstr`/`strchr` path check now runs against the original `endpoint` before `StripTrailingSlash`, so a lone trailing slash (`https://host/`) is recognized as an explicit mount path and returns `https://host` (templates then build `https://host/v1/...`).  A bare `https://host` (no slash) still gets `/api/catalog` appended — backward compatible.

Updated the test: `("https://host/", "https://host")` replaces the old expectation of `"https://host/api/catalog"`, with an updated comment explaining the root-mount semantics.  The double-slash edge-case note in the inline comment is updated to reflect that `https://host//` → path check fires → strip → `https://host/` → strip → `https://host` (two strips on separate calls would be needed, but StripTrailingSlash only strips once, so it returns `https://host/` — server-side 404, no silent wrong result, same as before).

**Comment by sfc-gh-rbachala** [comment]
Second round of fixes (`70d8aca`):

- **Build fix**: `RestCatalogOptions.host` rename to `baseUri` missed `pg_lake_table/src/transaction/track_iceberg_metadata_changes.c` (two `catalogOpts->host` sites at lines 446 and 793). Fixed.

- **Lint fix (round 2)**: pgindent reflowed two more comment blocks that I had tried to match by hand but got wrong:
  - `rest_catalog.c` inline `/*` block: "trailing slashes... resulting URL would / produce... so we / tolerate..."
  - `rest_catalog_auth.c` block: "treated... NOT passed / through... OAuth / token endpoint... (e.g. a / separate IdP)..."

  Applied pgindent's exact diffs from the CI log.

**Comment by sfc-gh-rbachala** [comment]
The remaining failures in run [#5](https://github.com/sfc-gh-rbachala/pg_lake/actions/runs/31617209060) are all the same transient runner issue you noted earlier — `FileNotFoundError: /tmp/pg_lake_tests/logfile` — PostgreSQL never came up in those runners.  Every failing job (`pgduck_server`, `pg_lake_copy`, `pg_lake_benchmark`, `test-isolation`, `test-upgrade`, `test-installcheck-postgres`) has the identical root cause and touches no REST catalog code.

Our tests pass:
- `test-check (17, check-pg_lake_iceberg)` ✅
- `test-check (18, check-pg_lake_iceberg)` ✅
- All build jobs ✅
- `lint-check-18` ✅

Could you re-run the failed jobs when you get a chance?

**Comment by sfc-gh-mslot** [comment]
ran CI for this again on a branch off current main (`marcoslot/ci-pr-507b`, 885dd57), since the fork PR doesn't get the pytest workflow. all 63 jobs are green this time, including lint. the new base-uri cases pass on 17 and 18, and `test_rest_endpoint_path_is_honored` passes on 16, 17 and 18 against the live Polaris fixture.

the root-mount fix does what it says, and `host` -> `baseUri` makes this a lot easier to follow.

one thing I probably should have called out when I suggested hoisting the path check above the strip: it makes the trailing slash significant, so `https://host` and `https://host/` now point at two different catalogs. that's a little surprising for a base URL, where most code treats those two as the same thing. before this PR a bare host with a trailing slash built `https://host//api/catalog/v1/...`, so my guess is it was already broken and nobody has a working config whose meaning silently changes here, but did you check that? if some server does tolerate the double slash, then those setups quietly start talking to `/v1/` instead.

a related consequence is that resolution is no longer idempotent for the root case. `https://host/` resolves to `https://host`, and resolving that again appends `/api/catalog`. it's fine as written because both builders resolve from the stored option value on every build, so maybe just something to keep in mind if we ever cache resolved options or write the normalized value back into the server option.

small thing in the new comment: it says `https://host//` becomes `https://host/` after one strip and then "further reduces to https://host", but `StripTrailingSlash` only takes off a single slash, so we stop at `https://host/` and emit `https://host//v1/...`. the conclusion is the same either way since the URL 404s, it's just that the intermediate step described isn't what happens. probably worth adding that input to `BASE_URI_CASES` so the behaviour is pinned instead of described.

and for the record the root-mounted path has unit coverage only, since none of the fixture catalogs is mounted at `/`. that seems acceptable to me, the reference `iceberg-rest-fixture` image would be the only way to get a real one and that is a much bigger change.

nothing else from me, looks good to merge once the trailing slash question is settled.

**Comment by sfc-gh-rbachala** [comment]
Good catches, addressed in `06de47b`.

**Comment bug fixed**
The "further reduces to `https://host`" step was wrong — `StripTrailingSlash` only removes one slash, so `https://host//` stops at `https://host/` and the template emits `https://host//v1/...`.  Updated the comment to say exactly that.

**`https://host//` test case added**
`("https://host//", "https://host/")` now pins the behavior in `BASE_URI_CASES` with a comment explaining the double-slash consequence.

**Trailing slash question**
Before this PR, `https://host/` and `https://host` were identical: the slash was stripped before the path check, so both resolved to `https://host/api/catalog`.  The change makes them semantically different — `https://host/` now means "root mount" and `https://host` keeps the legacy Polaris prefix.

For a silent breakage to occur, a user would need `rest_endpoint = 'https://polaris.example.com/'` (trailing slash, pointing at Polaris) AND their Polaris deployment would need to serve the Iceberg spec at `/v1/` directly (root mount) rather than at `/api/catalog/v1/`.  That combination seems effectively impossible: Polaris doesn't mount at root, so the request to `/v1/config` would 404 and the user would get an explicit error rather than a silently wrong result.

The non-idempotency note is a good one to keep in mind for caching — noted.

File: pg_lake_iceberg/src/rest_catalog/rest_catalog.c [hunk 1]
```diff
@@ -47,6 +48,62 @@ int			RestCatalogAuthType = REST_CATALOG_AUTH_TYPE_OAUTH2;
...
+	const char *schemeSep = strstr(endpoint, "://");
+	const char *authority = schemeSep ? schemeSep + 3 : endpoint;
+	bool		hasPath = (strchr(authority, '/') != NULL);
+	char	   *baseUri = StripTrailingSlash(pstrdup(endpoint), true);
+
+	if (baseUri[0] == '\0')
+		return baseUri;
+
+	if (!hasPath)
+		baseUri = psprintf("%s/api/catalog", baseUri);
```
Comment (sfc-gh-mslot) [inline pg_lake_iceberg/src/rest_catalog/rest_catalog.c]: thinking more about this as a human, this feels a bit too subtle

can we have something like `rest_catalog_host_path` that mirrors `rest_catalog_oauth_host_path` and defaults to /api/catalog?

**Comment by richiebachala** [comment]
Agreed — the trailing-slash distinction is too subtle. Here is what I'm planning to implement, please say if any of this is off:

**`rest_catalog_host_path` server option + GUC**

Mirrors `rest_catalog_oauth_host_path` / `oauth_endpoint` exactly:
- GUC `pg_lake_iceberg.rest_catalog_host_path` (default `NULL`)
- Server option `rest_catalog_host_path`
- New `char *hostPath` field on `RestCatalogOptions`

**Behavior in `ResolveRestCatalogBaseUri(endpoint, hostPath)`**

- `hostPath` is NULL → current auto-detect logic (strip trailing slash, then check for explicit path after authority, fall back to `/api/catalog` for bare hosts) — the root-mount-from-trailing-slash change is reverted
- `hostPath` is set → `baseUri = stripped_endpoint + hostPath`, no path detection at all
  - `hostPath = '/catalog'` → Lakekeeper
  - `hostPath = '/iceberg'` → Nessie  
  - `hostPath = ''` (empty string) → root mount
  - `hostPath = '/api/catalog'` → explicit Polaris (same result as bare endpoint today)

**Backward compat**
All existing bare-host Polaris configs work unchanged (auto-detect → `/api/catalog`).  Existing explicit-path configs (`rest_endpoint = 'https://host/catalog'`) also still work via auto-detect.  New catalogs use `rest_catalog_host_path` instead of embedding the path in `rest_endpoint`.

**Files touched**
`rest_catalog.h`, `rest_catalog.c`, `rest_catalog_options.c`, `init.c`, the test wrapper (add `host_path TEXT DEFAULT NULL` second arg), `test_rest_catalog_base_uri.py` (revert `https://host/` expectation, add `rest_catalog_host_path` cases).

Does this match what you had in mind?  Happy to proceed once confirmed.

**Comment by sfc-gh-mslot** [comment]
Seems I misread and the rest_endpoint and oauth_endpoint server properties are already inconsistent, so including the path in rest_endpoint is more consistent.

wonder how hard it would be to fix existing server properties in the 3.4--3.5 upgrades script to add /api/catalog to the configuration. Alternatively, maybe we should slightly break backwards compatibility, since it's pretty new and still undocumented.

**Comment by richiebachala** [comment]
The upgrade script approach is definitely doable. Here is what it would look like for each surface:

**User-created servers (`rest_endpoint` server option)**
The 3.4→3.5 script can iterate over all `iceberg_catalog` servers and append `/api/catalog` to any bare-host `rest_endpoint`:

```sql
DO $$
DECLARE
  srv record;
  endpoint text;
BEGIN
  FOR srv IN
    SELECT s.oid, s.srvname
    FROM pg_foreign_server s
    JOIN pg_foreign_data_wrapper w ON s.srvfdw = w.oid
    WHERE w.fdwname = 'iceberg_catalog'
      AND s.srvname NOT IN ('pg_lake_postgres_catalog',
                            'pg_lake_object_store_catalog',
                            'pg_lake_rest_catalog')
  LOOP
    SELECT option_value INTO endpoint
    FROM pg_options_to_table(
      (SELECT srvoptions FROM pg_foreign_server WHERE oid = srv.oid))
    WHERE option_name = 'rest_endpoint';

    -- bare host: no '/' after the authority
    IF endpoint IS NOT NULL
       AND regexp_match(endpoint, '^([a-z]+://)?[^/]+$') IS NOT NULL THEN
      EXECUTE format('ALTER SERVER %I OPTIONS (SET rest_endpoint %L)',
                     srv.srvname, endpoint || '/api/catalog');
    END IF;
  END LOOP;
END $$;
```

**Built-in catalog (`pg_lake_iceberg.rest_catalog_host` GUC)**
The built-in `pg_lake_rest_catalog` server has no `rest_endpoint` option — it reads from the GUC.  The GUC code default can change from `http://localhost:8181` to `http://localhost:8181/api/catalog`, which covers fresh installs.  For existing installs that explicitly set the GUC, the upgrade script can do:

```sql
DO $$
DECLARE v text;
BEGIN
  v := current_setting('pg_lake_iceberg.rest_catalog_host', true);
  IF v IS NOT NULL AND v <> ''
     AND regexp_match(v, '^([a-z]+://)?[^/]+$') IS NOT NULL THEN
    PERFORM set_config('pg_lake_iceberg.rest_catalog_host',
                       v || '/api/catalog', false);
    EXECUTE format('ALTER SYSTEM SET "pg_lake_iceberg.rest_catalog_host" TO %L',
                   v || '/api/catalog');
  END IF;
END $$;
```

`ALTER SYSTEM` in an upgrade script is unusual but functional — it writes `postgresql.auto.conf` and takes effect after the next `pg_reload_conf()` / restart.

**`ResolveRestCatalogBaseUri` simplification**
With the upgrade handling backward compat, the function becomes trivial:

```c
char *
ResolveRestCatalogBaseUri(const char *endpoint)
{
    if (endpoint == NULL)
        return NULL;
    return StripTrailingSlash(pstrdup(endpoint), true);
}
```

And we can add a validation step in `ValidateRestCatalogOptions` that rejects bare-host endpoints (no path after authority) with a clear error pointing users to update their `rest_endpoint`.

**My recommendation**
Go with the upgrade script — it is ~30 lines total and covers both surfaces cleanly.  The `bare` integration test parametrization in `test_rest_endpoint_path_is_honored` becomes obsolete and can be removed; the `explicit` variant (which is the only valid style going forward) stays.

Happy to implement all of this if you're aligned.

**Comment by richiebachala** [comment]
The upgrade script approach is definitely doable. Here is what it would look like for each surface:

**User-created servers (`rest_endpoint` server option)**
The 3.4→3.5 script can iterate over all `iceberg_catalog` servers and append `/api/catalog` to any bare-host `rest_endpoint`:

```sql
DO $$
DECLARE
  srv record;
  endpoint text;
BEGIN
  FOR srv IN
    SELECT s.oid, s.srvname
    FROM pg_foreign_server s
    JOIN pg_foreign_data_wrapper w ON s.srvfdw = w.oid
    WHERE w.fdwname = 'iceberg_catalog'
      AND s.srvname NOT IN ('pg_lake_postgres_catalog',
                            'pg_lake_object_store_catalog',
                            'pg_lake_rest_catalog')
  LOOP
    SELECT option_value INTO endpoint
    FROM pg_options_to_table(
      (SELECT srvoptions FROM pg_foreign_server WHERE oid = srv.oid))
    WHERE option_name = 'rest_endpoint';
    IF endpoint IS NOT NULL
       AND regexp_match(endpoint, '^([a-z]+://)?[^/]+$') IS NOT NULL THEN
      EXECUTE format('ALTER SERVER %I OPTIONS (SET rest_endpoint %L)',
                     srv.srvname, endpoint || '/api/catalog');
    END IF;
  END LOOP;
END $$;
```

**Built-in catalog (`pg_lake_iceberg.rest_catalog_host` GUC)**
The built-in `pg_lake_rest_catalog` server has no `rest_endpoint` option — it reads from the GUC. The code default changes from `http://localhost:8181` to `http://localhost:8181/api/catalog` for fresh installs. For existing installs that explicitly set the GUC the upgrade script can detect and migrate the bare-host value.

`ALTER SYSTEM` in an upgrade script is unusual but functional — it writes `postgresql.auto.conf` and takes effect after the next reload/restart.

**`ResolveRestCatalogBaseUri` simplification**
With the upgrade handling backward compat the function becomes trivial: just strip the trailing slash and return verbatim. We can also add a `ValidateRestCatalogOptions` guard that rejects bare-host endpoints with a clear error.

**My recommendation**
Go with the upgrade script — it is ~30 lines total and covers both surfaces cleanly. The `bare` parametrization of `test_rest_endpoint_path_is_honored` becomes obsolete (can be removed); the `explicit` variant stays as the only valid style going forward.

Happy to implement all of this if you are aligned.

**Comment by richiebachala** [comment]
The upgrade script approach is definitely doable. Here is what it would look like for each surface.

**User-created servers**
The 3.4 to 3.5 script iterates over all iceberg_catalog servers and appends /api/catalog to any bare-host rest_endpoint (no path after authority). The DO block is about 20 lines - finds each server, checks the endpoint with a regex, and runs ALTER SERVER ... OPTIONS (SET rest_endpoint ...) to patch it in place.

**Built-in catalog (pg_lake_iceberg.rest_catalog_host GUC)**
The built-in pg_lake_rest_catalog server has no rest_endpoint option - it reads from the GUC. The code default changes from http://localhost:8181 to http://localhost:8181/api/catalog for fresh installs. For existing installs that explicitly set the GUC, the upgrade script can detect and migrate the bare-host value. ALTER SYSTEM in an upgrade script is unusual but works - it writes postgresql.auto.conf and takes effect after the next reload.

**ResolveRestCatalogBaseUri simplification**
With the upgrade handling backward compat, the function becomes trivial: strip one trailing slash and return verbatim. We can also add a ValidateRestCatalogOptions guard that rejects bare-host endpoints (no path after authority) with a clear error pointing users to update their rest_endpoint.

**My recommendation**
Go with the upgrade script - it is about 30 lines total and covers both surfaces cleanly. The bare parametrization of test_rest_endpoint_path_is_honored becomes obsolete and can be removed; the explicit variant stays as the only valid style going forward.

Happy to implement all of this if you are aligned.

**Comment by sfc-gh-mslot** [comment]
> ALTER SYSTEM in an upgrade script is unusual but works - it writes postgresql.auto.conf and takes effect after the next reload.

I'd not go that route, it's blocked on some systems

There are probably not many users that have used CREATE SERVER in production already, so the bigger concern is the GUCs, but not much we can do about those.

We could have a rest_endpoint_path or something like that, but if we default to /api/catalog we'll forever annoy anyone not using Polaris.

I'm now leaning towards just declaring this a breaking change in the 3.5 release notes.

I'm not sure how much the exception for the bare host path adds, since some users will have used `https://xxx/polaris`. Maybe remove that.

**Comment by richiebachala** [comment]
Understood — breaking change in release notes it is.

Plan:
- `ResolveRestCatalogBaseUri` = strip one trailing slash, return verbatim. No path detection, no `/api/catalog` appending.
- GUC default changes from `http://localhost:8181` to `http://localhost:8181/api/catalog`. Existing installs with an explicit bare-host GUC need a manual update; the release notes document this.
- User-created servers: a lightweight DO block in `pg_lake_iceberg--3.4--3.5.sql` (no `ALTER SYSTEM`) patches bare-host `rest_endpoint` server options to add `/api/catalog`. Safe to include, handles the few production CREATE SERVER configs that exist, and can be skipped without loss if you prefer a clean break there too.
- No bare-host validation guard — agreed that it adds marginal value and would just be noise.
- `test_rest_endpoint_path_is_honored[bare]` removed; `[explicit]` stays.
- Tests that set the GUC to a bare scheme-less host get updated to include `/api/catalog`.

Does the user-created server DO block belong in the upgrade script or do you want a fully clean break on both surfaces?

**Comment by sfc-gh-mslot** [comment]
> Does the user-created server DO block belong in the upgrade script or do you want a fully clean break on both surfaces?

we can break both

**Comment by sfc-gh-rbachala** [comment]
Done — `8668bf8` implements the clean break:

- `ResolveRestCatalogBaseUri` now just strips one trailing slash and returns verbatim. All path detection and the `/api/catalog` fallback are gone.
- GUC default: `http://localhost:8181` → `http://localhost:8181/api/catalog`.
- `test_rest_endpoint_path_is_honored[bare]` removed; `[explicit]` stays.
- All four test fixtures that set `rest_catalog_host` to a bare scheme-less host updated to include `/api/catalog`.
- Unit test expected values updated (bare hosts no longer get `/api/catalog` appended).

**Comment by sfc-gh-rbachala** [comment]
Done -- 8668bf8 implements the clean break. ResolveRestCatalogBaseUri now just strips one trailing slash and returns verbatim -- all path detection and the /api/catalog fallback are gone. GUC default changes from http://localhost:8181 to http://localhost:8181/api/catalog. test_rest_endpoint_path_is_honored[bare] removed, [explicit] stays. All four test fixtures that set rest_catalog_host to a bare scheme-less host updated to include /api/catalog. Unit test expected values updated.

**Comment by sfc-gh-rbachala** [comment]
Done -- 8668bf8 implements the clean break:

- ResolveRestCatalogBaseUri now just strips one trailing slash and returns verbatim -- all path detection and the /api/catalog fallback are gone.
- GUC default: http://localhost:8181 -> http://localhost:8181/api/catalog.
- test_rest_endpoint_path_is_honored[bare] removed; [explicit] stays.
- All four test fixtures that set rest_catalog_host to a bare scheme-less host updated to include /api/catalog.
- Unit test expected values updated (bare hosts no longer get /api/catalog appended).

File: test_common/helpers/polaris.py [hunk 2]
```diff
@@ -219,7 +219,7 @@ def set_polaris_gucs(
 
         run_command_outside_tx(
             [
-                f"""ALTER SYSTEM SET pg_lake_iceberg.rest_catalog_host TO '{server_params.POLARIS_HOSTNAME}:{server_params.POLARIS_PORT}'""",
+                f"""ALTER SYSTEM SET pg_lake_iceberg.rest_catalog_host TO '{server_params.POLARIS_HOSTNAME}:{server_params.POLARIS_PORT}/api/catalog'""",
```
Comment (sfc-gh-mslot) [inline test_common/helpers/polaris.py]: The four GUC fixtures got `/api/catalog`, but the `CREATE SERVER ... rest_endpoint` sites did not. 21 of them still build `endpoint = f"http://localhost:{server_params.POLARIS_PORT}"`: 20 in `test_modify_iceberg_rest_table.py` plus the shared `create_iceberg_user_server_rest_table` fixture in `test_writable_iceberg_common.py`. Server options beat the GUC, so those resolve to `/v1/oauth/tokens` and Polaris 404s.

File: pg_lake_table/tests/pytests/test_modify_iceberg_rest_table.py [hunk 3]
```diff
@@ -732,6 +732,89 @@ def test_server_option_overrides_guc(
         superuser_conn.commit()
 
 
+# Only the explicit endpoint style is tested: rest_endpoint must include
+# the full catalog mount path (e.g. "/api/catalog" for Polaris).
+@pytest.mark.parametrize("endpoint_style", ["explicit"])
```
Comment (sfc-gh-mslot) [inline pg_lake_table/tests/pytests/test_modify_iceberg_rest_table.py]: nit: single-value parametrize now, and the `else host` two lines below is dead. Worth dropping both since `[bare]` is gone.

**Review by sfc-gh-mslot — COMMENTED** [review]
Ran this on a branch off current main since the fork PR skips pytest_all: run 32465586424. `test-check-pg-lake-table (17, check-pg_lake_table, 4)` fails with `Rest Catalog OAuth token request failed (HTTP 404)` on `CREATE TABLE` against both `crud_test_server` and `rest_opt_override_rest_endpoint`. Details inline. The pg16 and pg18 failures in that run are the usual missing `/tmp/pg_lake_tests/logfile` setup problem, main is green at the same base.

The C side reads right to me. Nothing uses `->host` any more, both builders normalize, and every resolution path goes through one of them. But the description now says the opposite of what the code does, it still promises bare-host Polaris configs keep working with no user action, and the 3.5 release note gets written off that. So not ready to merge until the fixtures are migrated and the description is rewritten.

nit: the `rest_endpoint 'https://polaris.example.com'` example in the 3.3--3.4 script comment is a broken config now.

arcticowl's SSRF finding looks pre-existing to me, `rest_endpoint` was already fully user-controlled including the host.

**Comment by sfc-gh-rbachala** [comment]
Addressed in `84c6f49` (rebased on current main):

**Missed server-option sites (the 404s)**
21 `endpoint = f"http://localhost:{PORT}"` assignments were not updated -- the four GUC fixtures got `/api/catalog` but the `CREATE SERVER ... rest_endpoint` sites did not, and server options beat the GUC. Fixed: 20 in `test_modify_iceberg_rest_table.py` and the shared `create_iceberg_user_server_rest_table` fixture in `test_writable_iceberg_common.py`.

**Parametrize nit**
`@pytest.mark.parametrize("endpoint_style", ["explicit"])` removed entirely; function signature cleaned up; dead `host` variable and `else host` branch dropped.

**PR description**
Rewritten to reflect the breaking change; "bare-host configs keep working with no user action" removed.

**3.3--3.4 upgrade script example**
`rest_endpoint 'https://polaris.example.com'` updated to `https://polaris.example.com/api/catalog`.

**Merge conflict**
Rebased cleanly on `7f06ebd`; `rest_catalog_ops.c` and `src/test/rest_catalog.c` (which gained `get_rest_metadata_location` / `get_rest_vended_credentials` from the vended-creds PR) both resolved -- all four test functions coexist in the file.

**Comment by sfc-gh-rbachala** [comment]
260dc63 fixes two issues found in the post-push review:

1. The 21 endpoint = f"http://localhost:{PORT}" fixes in test_modify_iceberg_rest_table.py were lost during the rebase conflict resolution. Re-applied: all 21 sites now use /api/catalog.

2. The parametrize + dead-code cleanup (removing @pytest.mark.parametrize(["explicit"]), host variable, else host branch) was also lost. Re-applied: test_rest_endpoint_path_is_honored is now a plain function with a hardcoded endpoint.

3. The pg_lake_iceberg--3.3--3.4.sql comment example was not in the rebased commit. Fixed: https://polaris.example.com -> https://polaris.example.com/api/catalog.

**Comment by sfc-gh-rbachala** [comment]
Run 32708045644 on commit 260dc63: all builds green, lint green, test-check (17, check-pg_lake_iceberg) green. All pg18 failures are the same FileNotFoundError: /tmp/pg_lake_tests/logfile runner infrastructure issue -- PostgreSQL never came up in those runners. Same pattern as before, same job passes on main.

**Comment by sfc-gh-rbachala** [comment]
Rebased on 94413e9 (current main). The 79cef0c fixture refactor conflicts resolved:

- All three test files (polaris.py, test_modify_iceberg_rest_table.py, test_polaris_catalog_writable.py) take your apply_polaris_gucs()/reset_polaris_gucs() structure.
- The host variable inside apply_polaris_gucs updated to include /api/catalog:
  host = f"{POLARIS_HOSTNAME}:{POLARIS_PORT}/api/catalog"
  so the new wait_for_reloaded_settings() check also waits for the correct value.

All validations pass: no bare-host endpoints, no opts->host refs, GUC default and upgrade script both correct. Commit: 889ee9b

**Comment by sfc-gh-mslot** [comment]
I ran CI for this on `marcoslot/pr507-rest-catalog-base-uri` (your commit as-is, no rebase) and it came back red on PG 16 and 18. `init.c` now has `/api/catalog` in the `rest_catalog_host` boot value, but `rest_catalog.c` still has `RestCatalogHost = "http://localhost:8181"`, and assert builds check that the two agree:

```
LOG:  GUC (PGC_STRING) pg_lake_iceberg.rest_catalog_host, boot_val=http://localhost:8181/api/catalog, C-var=http://localhost:8181
TRAP: failed Assert("check_GUC_init(variable)"), File: "guc.c", Line: 4944, PID: 476
```

Postgres dies while preloading pg_lake_iceberg, so every test fails as `FileNotFoundError: /tmp/pg_lake_tests/logfile` before it runs. PG 17 stayed green only because that image is not built with asserts, which makes the failure look version-specific when it isn't.

With that fixed, `test_polaris_server_option_enables_vending_with_guc_off` still failed on both versions. It creates its own server instead of going through `apply_polaris_gucs`, so its `rest_endpoint` was still a bare `host:port`, and the OAuth request went to `host:port/v1/oauth/tokens` which Polaris answers with a 404.

Both fixes are on that branch now and CI is green there, 63/63: https://github.com/Snowflake-Labs/pg_lake/actions/runs/32770163225

Can you cherry-pick them onto this branch so we can merge?

```
git fetch https://github.com/Snowflake-Labs/pg_lake marcoslot/pr507-rest-catalog-base-uri
git cherry-pick eb3d5e2df1568702c8666335edb43659339b68c1 d493c21290ece25f584bf9d7ebcccf5a374dfbb6
```

**Comment by sfc-gh-rbachala** [comment]
Fixed in 144526a: rest_catalog.c still had the old C-variable initializer:

  char *RestCatalogHost = "http://localhost:8181";

Updated to match init.c boot_val:

  char *RestCatalogHost = "http://localhost:8181/api/catalog";

Thank you for catching that.

**Comment by sfc-gh-rbachala** [comment]
Also found and fixed 9c3d486: test_vended_credentials.py (added in 7f06ebd) has 5 mock-server GUC sets at http://127.0.0.1:{port} without /api/catalog. Same pattern as test_token_refresh. All five updated. Full scan now shows no remaining bare-host rest_catalog_host SET lines anywhere in the test suite.

**Comment by sfc-gh-rbachala** [comment]
36e78e9: one more -- test_vended_credentials_polaris.py line 311 had endpoint = f"http://{POLARIS_HOSTNAME}:{POLARIS_PORT}" for a CREATE SERVER rest_endpoint. Same 404 pattern. Fixed. Full repo scan is now clean: no remaining bare-host endpoint assignments or bare ALTER SYSTEM SET rest_catalog_host anywhere in the test suite.

**Review by sfc-gh-mslot — APPROVED** [review]
